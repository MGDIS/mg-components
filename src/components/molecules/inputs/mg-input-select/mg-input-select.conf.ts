/**
 * type SelectOption
 * use to match select>option attributes
 */
export type SelectOption = {
  title: string;
  value: any;
  disabled?: boolean;
  group?: string;
};

/**
 * type OptGroup
 * use to match select>optgroup attributes
 */
export type OptGroup = {
  group: string;
  options: SelectOption[];
};
