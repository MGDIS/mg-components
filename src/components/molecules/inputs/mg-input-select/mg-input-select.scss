@use '../styles/mg-input.scss';
@use '../styles/mg-input-box.scss';
@use '../styles/mg-input-width.scss';
@use '../../../../styles/a11y.scss';

.mg-input {
  &__box {
    width: 100%;
    padding-top: 0;
    padding-bottom: 0;
    padding-right: 1rem;
  }
  &.mg-input--select:not(.mg-input--label-on-top) .mg-input__box {
    max-width: unset;
  }
}
