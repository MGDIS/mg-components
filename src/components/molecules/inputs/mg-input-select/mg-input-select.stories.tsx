import { h } from "@stencil/core";

export default {
  component: 'mg-input-select',
  title: 'Molecules/Inputs/mg-input-select',
};

/**
 * 1. camelCase arguments must be written in the template, for exemple labelOnTop must be placed in the template as label-on-top={args.labelOnTop}
 * 2. boolean arguments with a default true value must be added like display-character-left={args.displayCharacterLeft ? 'true' : 'false'}
 */
 const Template = args => {
  const labelOnTop = args.labelOnTop;
  delete args.labelOnTop;
  const labelHide = args.labelHide;
  delete args.labelHide;
  const helpText = args.helpText;
  delete args.helpText;
  // return element
  return <mg-input-select
    {...args}
    label-on-top={labelOnTop}
    label-hide={labelHide}
    help-text={helpText}
  ></mg-input-select>
}

export const MgInputSelect = Template.bind({});
MgInputSelect.args = {
  // Global
  value: '',
  identifier: 'identifier',
  name: 'input-name',
  // Label
  label: 'Label',
  labelOnTop: false,
  labelHide: false,
  // Input
  required: true,
  readonly: false,
  disabled: false,
  items: ['blu', 'bli', 'blo', 'bla'],
  // Tooltip
  tooltip: 'This is a tooltip',
  // Help Text
  helpText: 'Help text with html <strong>bold</strong>, <em>italic</em>.',
};

/**
 * Using object
 */

export const WithObjectItems = Template.bind({});
WithObjectItems.args = {
  ...MgInputSelect.args,
  // remove feature to focus on pattern
  tooltip: '',
  helpText: '',
  required: false,
  //
  items: [
    { title: 'blu', value: 'blublu' },
    { title: 'bli', value: 'blibli' },
    { title: 'blo', value: 'bloblo' },
    { title: 'bla', value: 'blabla', disabled: true },
  ],
};

/**
 * Using group object
 */

export const WithGroups = Template.bind({});
WithGroups.args = {
  ...WithObjectItems.args,
  items: [
    { title: 'blu', value: 'blublu', group: '1st group' },
    { title: 'bli', value: 'blibli', group: '2nd group' },
    { title: 'blo', value: 'bloblo', group: '1st group' },
    { title: 'bla', value: 'blabla' },
  ],
};
