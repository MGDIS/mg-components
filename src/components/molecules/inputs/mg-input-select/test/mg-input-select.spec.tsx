import { h } from '@stencil/core';
import { newSpecPage } from '@stencil/core/testing';
import { MgInputSelect } from '../mg-input-select';
import { SelectOption } from '../mg-input-select.conf';
import messages from '../../../../../locales/en/messages.json';

const getPage = async args => {
  const page = newSpecPage({
    components: [MgInputSelect],
    template: () => <mg-input-select {...args}></mg-input-select>,
  });

  jest.runAllTimers();

  return page;
};

describe('mg-input-select', () => {
  beforeEach(() => jest.useFakeTimers());
  afterEach(() => jest.runOnlyPendingTimers());
  test.each([
    { label: 'label', identifier: 'identifier', items: ['blu', 'bli', 'blo', 'bla'] },
    { label: 'label', identifier: 'identifier', items: ['blu', 'bli', 'blo', 'bla'], labelHide: true },
    {
      label: 'label',
      identifier: 'identifier',
      items: [
        { title: 'blu', value: 'u' },
        { title: 'bli', value: 'i' },
        { title: 'blo', value: 'o' },
        { title: 'bla', value: 'a' },
      ],
    },
    {
      label: 'label',
      identifier: 'identifier',
      items: [
        { title: 'blu', value: 'u', group: 'blu' },
        { title: 'bli', value: 'i', group: 'blu' },
        { title: 'blo', value: 'o', group: 'blo' },
        { title: 'bla', value: 'a' },
      ],
    },
    { label: 'label', identifier: 'identifier', items: ['blu', 'bli', 'blo', 'bla'], labelOnTop: true },
    { label: 'label', identifier: 'identifier', items: ['blu', 'bli', 'blo', 'bla'], readonly: true },
    { label: 'label', identifier: 'identifier', items: ['blu', 'bli', 'blo', 'bla'], readonly: true, labelOnTop: true, tooltip: 'Tooltip message' },
    { label: 'label', identifier: 'identifier', items: ['blu', 'bli', 'blo', 'bla'], tooltip: 'My Tooltip Message' },
    { label: 'label', identifier: 'identifier', items: ['blu', 'bli', 'blo', 'bla'], tooltip: 'My Tooltip Message', labelOnTop: true },
    { label: 'label', identifier: 'identifier', items: ['blu', 'bli', 'blo', 'bla'], placeholder: 'placeholder' },
    { label: 'label', identifier: 'identifier', items: ['blu', 'bli', 'blo', 'bla'], placeholder: 'placeholder', placeholderHide: true },
    { label: 'label', identifier: 'identifier', items: ['blu', 'bli', 'blo', 'bla'], placeholder: 'placeholder', placeholderDisabled: true },
    { label: 'label', identifier: 'identifier', items: ['blu', 'bli', 'blo', 'bla'], placeholder: 'placeholder', placeholderDisabled: true, value: 'ble' },
  ])('Should render with args %s:', async args => {
    const { root } = await getPage(args);
    expect(root).toMatchSnapshot();
  });

  test.each(['', ' ', undefined])('Should not render with invalid identifier property: %s', async identifier => {
    expect.assertions(1);
    try {
      await getPage({ identifier, items: ['blu', 'bli', 'blo', 'bla'] });
    } catch (err) {
      expect(err.message).toMatch('<mg-input> prop "identifier" is required.');
    }
  });

  test.each(['', ' ', undefined])('Should throw error with invalid label property : %s', async label => {
    expect.assertions(1);
    try {
      await getPage({ identifier: 'identifier', label, items: ['blu', 'bli', 'blo', 'bla'] });
    } catch (err) {
      expect(err.message).toMatch('<mg-input> prop "label" is required.');
    }
  });

  test('Should throw an error with labelOnTop & labelHide set to true', async () => {
    expect.assertions(1);
    try {
      await getPage({ identifier: 'identifier', label: 'batman', labelOnTop: true, labelHide: true, items: ['batman', 'joker'] });
    } catch (err) {
      expect(err.message).toMatch('<mg-input> prop "labelOnTop" must not be paired with the prop "labelHide"');
    }
  });

  test.each([[['blu', { title: 'blu', value: 'blu' }]], [['blu', { blu: 'blu' }]], [[{ title: 'blu', value: 'blu' }, { blu: 'blu' }]]])(
    'Should throw error with invalid items property : %s',
    async items => {
      expect.assertions(1);
      try {
        await getPage({ label: 'Label', items });
      } catch (err) {
        expect(err.message).toMatch('<mg-input-select> prop "items" is required and all items must be the same type, string or Option.');
      }
    },
  );

  test.each([
    { items: ['batman', 'robin', 'joker', 'bane'], selectedOption: '' },
    { items: ['batman', 'robin', 'joker', 'bane'], selectedOption: 3 },
    {
      items: [
        { title: 'batman', value: 'u' },
        { title: 'robin', value: 'i' },
        { title: 'joker', value: 'o' },
        { title: 'bane', value: 'a' },
      ],
      selectedOption: 2,
    },
    {
      items: [
        { title: 'batman', value: 1 },
        { title: 'robin', value: 2 },
        { title: 'joker', value: 3 },
        { title: 'bane', value: 4 },
      ],
      selectedOption: 1,
    },
    {
      items: [
        { title: 'batman', value: true },
        { title: 'robin', value: false },
      ],
      selectedOption: 0,
    },
    {
      items: [
        { title: 'blu', value: undefined },
        { title: 'bli', value: 42 },
      ],
      selectedOption: 0,
    },
    {
      items: [
        { title: 'batman', value: { name: 'Batman' } },
        { title: 'robin', value: { name: 'Robin' } },
      ],
      selectedOption: 1,
    },
    {
      items: [
        { title: 'Batman', value: { name: 'Batman', group: 'Heroes' } },
        { title: 'Robin', value: { name: 'Robin', group: 'Heroes' } },
        { title: 'Joker', value: { name: 'Joker', group: 'Super Vilains' } },
        { title: 'Bane', value: { name: 'Bane', group: 'Super Vilains' } },
      ],
      selectedOption: 2,
    },
  ])('Should trigger events for (%s)', async ({ items, selectedOption }) => {
    const args = { label: 'label', items, identifier: 'identifier', helpText: 'My help text', value: 'blu' };
    const page = await getPage(args);

    const element = page.doc.querySelector('mg-input-select');
    const input = element.shadowRoot.querySelector('select');

    //mock validity
    input.checkValidity = jest.fn(() => true);
    Object.defineProperty(input, 'validity', {
      get: jest.fn(() => ({
        valueMissing: false,
      })),
    });

    jest.spyOn(page.rootInstance.valueChange, 'emit');

    input.dispatchEvent(new CustomEvent('focus', { bubbles: true }));
    await page.waitForChanges();

    expect(page.root).toMatchSnapshot(); //Snapshot on focus

    input.value = items[selectedOption]?.title || items[selectedOption] || selectedOption;
    input.dispatchEvent(new CustomEvent('input', { bubbles: true }));
    await page.waitForChanges();
    const expectedEmitValue = selectedOption !== '' ? (typeof items[selectedOption] === 'object' ? (items[selectedOption] as SelectOption).value : items[selectedOption]) : null;
    expect(page.rootInstance.valueChange.emit).toHaveBeenCalledWith(expectedEmitValue);
  });

  describe.each(['readonly', 'disabled'])('validity, case next state is %s', nextState => {
    test.each([
      { validity: true, valueMissing: false },
      { validity: false, valueMissing: true },
    ])('validity (%s), valueMissing (%s)', async ({ validity, valueMissing }) => {
      const args = { label: 'label', items: ['blu', 'bli', 'blo', 'bla'], identifier: 'identifier', patternErrorMessage: 'Non' };
      const page = await getPage(args);

      const element = page.doc.querySelector('mg-input-select');
      const input = element.shadowRoot.querySelector('select');

      //mock validity
      input.checkValidity = jest.fn(() => validity);
      Object.defineProperty(input, 'validity', {
        get: jest.fn(() => ({
          valueMissing,
        })),
      });

      input.dispatchEvent(new CustomEvent('blur', { bubbles: true }));
      await page.waitForChanges();

      if (validity) {
        expect(page.rootInstance.errorMessage).toBeUndefined();
      } else if (valueMissing) {
        expect(page.rootInstance.errorMessage).toEqual(messages.errors.required);
      }
      expect(page.rootInstance.valid).toEqual(validity);
      expect(page.rootInstance.invalid).toEqual(!validity);
      if (valueMissing) {
        expect(page.root).toMatchSnapshot(); //Snapshot with readonly/disabled FALSE
        element[nextState] = true;
        await page.waitForChanges();
        expect(page.root).toMatchSnapshot(); //Snapshot with readonly/disabled TRUE
      }
    });
  });

  test("display error with displayError component's public method", async () => {
    const page = await getPage({ label: 'label', items: ['batman', 'robin', 'joker', 'bane'], identifier: 'identifier', required: true });

    expect(page.root).toMatchSnapshot();

    const element = page.doc.querySelector('mg-input-select');
    const input = element.shadowRoot.querySelector('select');

    //mock validity
    input.checkValidity = jest.fn(() => false);
    Object.defineProperty(input, 'validity', {
      get: jest.fn(() => ({
        valueMissing: true,
      })),
    });

    await element.displayError();

    await page.waitForChanges();

    expect(page.root).toMatchSnapshot();
  });

  test.each(['fr', 'xx'])('display error message with locale: %s', async lang => {
    const page = await getPage({ label: 'label', items: ['batman', 'robin', 'joker', 'bane'], identifier: 'identifier', required: true, lang });
    const element = page.doc.querySelector('mg-input-select');
    const input = element.shadowRoot.querySelector('select');

    //mock validity
    input.checkValidity = jest.fn(() => false);
    Object.defineProperty(input, 'validity', {
      get: jest.fn(() => ({
        valueMissing: true,
      })),
    });

    await element.displayError();

    await page.waitForChanges();

    expect(page.root).toMatchSnapshot();
  });

  test('Should remove error on input', async () => {
    const page = await getPage({ label: 'label', items: ['batman', 'robin', 'joker', 'bane'], identifier: 'identifier', required: true });
    const element = page.doc.querySelector('mg-input-select');
    const input = element.shadowRoot.querySelector('select');

    //mock validity
    input.checkValidity = jest.fn().mockReturnValueOnce(false).mockReturnValueOnce(true).mockReturnValueOnce(true);
    Object.defineProperty(input, 'validity', {
      get: jest
        .fn()
        .mockReturnValueOnce({
          valueMissing: true,
        })
        .mockReturnValueOnce({
          valueMissing: true,
        })
        .mockReturnValueOnce({
          valueMissing: false,
        })
        .mockReturnValueOnce({
          valueMissing: false,
        }),
    });

    await element.displayError();

    await page.waitForChanges();

    expect(page.rootInstance.hasDisplayedError).toEqual(true);
    expect(page.rootInstance.errorMessage).toEqual(messages.errors.required);

    input.value = 'batman';
    input.dispatchEvent(new CustomEvent('input', { bubbles: true }));
    await page.waitForChanges();

    expect(page.rootInstance.hasDisplayedError).toEqual(true);
    expect(page.rootInstance.errorMessage).toBeUndefined();

    input.dispatchEvent(new CustomEvent('blur', { bubbles: true }));
    await page.waitForChanges();

    expect(page.rootInstance.hasDisplayedError).toEqual(false);
  });

  test('Should remove error on input when required change dynamically', async () => {
    const page = await getPage({ label: 'label', items: ['batman', 'robin', 'joker', 'bane'], identifier: 'identifier', required: true });
    const element = page.doc.querySelector('mg-input-select');
    const input = element.shadowRoot.querySelector('select');

    //mock validity
    input.checkValidity = jest.fn().mockReturnValueOnce(false).mockReturnValueOnce(true).mockReturnValueOnce(true);
    Object.defineProperty(input, 'validity', {
      get: jest
        .fn()
        .mockReturnValueOnce({
          valueMissing: true,
        })
        .mockReturnValueOnce({
          valueMissing: true,
        })
        .mockReturnValueOnce({
          valueMissing: false,
        })
        .mockReturnValueOnce({
          valueMissing: false,
        }),
    });

    await element.displayError();
    await page.waitForChanges();

    expect(page.rootInstance.hasDisplayedError).toEqual(true);
    expect(page.rootInstance.errorMessage).toEqual(messages.errors.required);

    element.required = false;
    await page.waitForChanges();

    // Error message should disapear and change the hasDisplayedError status
    expect(page.rootInstance.hasDisplayedError).toEqual(false);
    expect(page.rootInstance.errorMessage).toBeUndefined();

    element.required = true;
    await page.waitForChanges();

    // If back on required the message is still not displayed
    expect(page.rootInstance.hasDisplayedError).toEqual(false);
    expect(page.rootInstance.errorMessage).toBeUndefined();
  });
});
