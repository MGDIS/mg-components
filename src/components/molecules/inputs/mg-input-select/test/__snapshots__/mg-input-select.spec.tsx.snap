// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`mg-input-select Should render with args {
  label: 'label',
  identifier: 'identifier',
  items: [Array],
  labelHide: true
}: 1`] = `
<mg-input-select>
  <mock:shadow-root>
    <div class="mg-input mg-input--select">
      <mg-input-title class="sr-only" identifier="identifier">
        label
      </mg-input-title>
      <div class="mg-input__input-container">
        <div class="mg-input__input">
          <select aria-describedby="" class="mg-input__box" id="identifier" name="identifier" title="Veuillez sélectionner une valeur">
            <option value="">
              Veuillez sélectionner une valeur
            </option>
            <option value="0">
              blu
            </option>
            <option value="1">
              bli
            </option>
            <option value="2">
              blo
            </option>
            <option value="3">
              bla
            </option>
          </select>
        </div>
      </div>
    </div>
  </mock:shadow-root>
</mg-input-select>
`;

exports[`mg-input-select Should render with args {
  label: 'label',
  identifier: 'identifier',
  items: [Array],
  labelOnTop: true
}: 1`] = `
<mg-input-select>
  <mock:shadow-root>
    <div class="mg-input mg-input--label-on-top mg-input--select">
      <div class="mg-input__title">
        <mg-input-title identifier="identifier">
          label
        </mg-input-title>
      </div>
      <div class="mg-input__input-container">
        <div class="mg-input__input">
          <select aria-describedby="" class="mg-input__box" id="identifier" name="identifier" title="Veuillez sélectionner une valeur">
            <option value="">
              Veuillez sélectionner une valeur
            </option>
            <option value="0">
              blu
            </option>
            <option value="1">
              bli
            </option>
            <option value="2">
              blo
            </option>
            <option value="3">
              bla
            </option>
          </select>
        </div>
      </div>
    </div>
  </mock:shadow-root>
</mg-input-select>
`;

exports[`mg-input-select Should render with args {
  label: 'label',
  identifier: 'identifier',
  items: [Array],
  placeholder: 'placeholder'
}: 1`] = `
<mg-input-select>
  <mock:shadow-root>
    <div class="mg-input mg-input--select">
      <mg-input-title identifier="identifier">
        label
      </mg-input-title>
      <div class="mg-input__input-container">
        <div class="mg-input__input">
          <select aria-describedby="" class="mg-input__box" id="identifier" name="identifier" title="placeholder">
            <option value="">
              placeholder
            </option>
            <option value="0">
              blu
            </option>
            <option value="1">
              bli
            </option>
            <option value="2">
              blo
            </option>
            <option value="3">
              bla
            </option>
          </select>
        </div>
      </div>
    </div>
  </mock:shadow-root>
</mg-input-select>
`;

exports[`mg-input-select Should render with args {
  label: 'label',
  identifier: 'identifier',
  items: [Array],
  placeholder: 'placeholder',
  placeholderDisabled: true
}: 1`] = `
<mg-input-select>
  <mock:shadow-root>
    <div class="mg-input mg-input--select">
      <mg-input-title identifier="identifier">
        label
      </mg-input-title>
      <div class="mg-input__input-container">
        <div class="mg-input__input">
          <select aria-describedby="" class="mg-input__box" id="identifier" name="identifier" title="placeholder">
            <option value="">
              placeholder
            </option>
            <option value="0">
              blu
            </option>
            <option value="1">
              bli
            </option>
            <option value="2">
              blo
            </option>
            <option value="3">
              bla
            </option>
          </select>
        </div>
      </div>
    </div>
  </mock:shadow-root>
</mg-input-select>
`;

exports[`mg-input-select Should render with args {
  label: 'label',
  identifier: 'identifier',
  items: [Array],
  placeholder: 'placeholder',
  placeholderDisabled: true,
  value: 'ble'
}: 1`] = `
<mg-input-select value="ble">
  <mock:shadow-root>
    <div class="mg-input mg-input--select">
      <mg-input-title identifier="identifier">
        label
      </mg-input-title>
      <div class="mg-input__input-container">
        <div class="mg-input__input">
          <select aria-describedby="" class="mg-input__box" id="identifier" name="identifier" title="placeholder">
            <option value="">
              placeholder
            </option>
            <option value="0">
              blu
            </option>
            <option value="1">
              bli
            </option>
            <option value="2">
              blo
            </option>
            <option value="3">
              bla
            </option>
          </select>
        </div>
      </div>
    </div>
  </mock:shadow-root>
</mg-input-select>
`;

exports[`mg-input-select Should render with args {
  label: 'label',
  identifier: 'identifier',
  items: [Array],
  placeholder: 'placeholder',
  placeholderHide: true
}: 1`] = `
<mg-input-select>
  <mock:shadow-root>
    <div class="mg-input mg-input--select">
      <mg-input-title identifier="identifier">
        label
      </mg-input-title>
      <div class="mg-input__input-container">
        <div class="mg-input__input">
          <select aria-describedby="" class="mg-input__box" id="identifier" name="identifier" title="placeholder">
            <option value="">
              placeholder
            </option>
            <option value="0">
              blu
            </option>
            <option value="1">
              bli
            </option>
            <option value="2">
              blo
            </option>
            <option value="3">
              bla
            </option>
          </select>
        </div>
      </div>
    </div>
  </mock:shadow-root>
</mg-input-select>
`;

exports[`mg-input-select Should render with args {
  label: 'label',
  identifier: 'identifier',
  items: [Array],
  readonly: true
}: 1`] = `
<mg-input-select>
  <mock:shadow-root>
    <div class="mg-input mg-input--readonly mg-input--select">
      <mg-input-title identifier="identifier">
        label
      </mg-input-title>
      <div class="mg-input__input-container">
        <strong></strong>
      </div>
    </div>
  </mock:shadow-root>
</mg-input-select>
`;

exports[`mg-input-select Should render with args {
  label: 'label',
  identifier: 'identifier',
  items: [Array],
  readonly: true,
  labelOnTop: true,
  tooltip: 'Tooltip message'
}: 1`] = `
<mg-input-select>
  <mock:shadow-root>
    <div class="mg-input mg-input--label-on-top mg-input--readonly mg-input--select">
      <div class="mg-input__title">
        <mg-input-title identifier="identifier">
          label
        </mg-input-title>
      </div>
      <div class="mg-input__input-container">
        <strong></strong>
      </div>
    </div>
  </mock:shadow-root>
</mg-input-select>
`;

exports[`mg-input-select Should render with args {
  label: 'label',
  identifier: 'identifier',
  items: [Array],
  tooltip: 'My Tooltip Message'
}: 1`] = `
<mg-input-select>
  <mock:shadow-root>
    <div class="mg-input mg-input--select">
      <mg-input-title identifier="identifier">
        label
      </mg-input-title>
      <div class="mg-input__input-container">
        <div class="mg-input__input">
          <select aria-describedby="" class="mg-input__box" id="identifier" name="identifier" title="Veuillez sélectionner une valeur">
            <option value="">
              Veuillez sélectionner une valeur
            </option>
            <option value="0">
              blu
            </option>
            <option value="1">
              bli
            </option>
            <option value="2">
              blo
            </option>
            <option value="3">
              bla
            </option>
          </select>
          <mg-tooltip identifier="identifier-tooltip" message="My Tooltip Message">
            <mg-icon icon="info-circle"></mg-icon>
          </mg-tooltip>
        </div>
      </div>
    </div>
  </mock:shadow-root>
</mg-input-select>
`;

exports[`mg-input-select Should render with args {
  label: 'label',
  identifier: 'identifier',
  items: [Array],
  tooltip: 'My Tooltip Message',
  labelOnTop: true
}: 1`] = `
<mg-input-select>
  <mock:shadow-root>
    <div class="mg-input mg-input--label-on-top mg-input--select">
      <div class="mg-input__title">
        <mg-input-title identifier="identifier">
          label
        </mg-input-title>
        <mg-tooltip identifier="identifier-tooltip" message="My Tooltip Message">
          <mg-icon icon="info-circle"></mg-icon>
        </mg-tooltip>
      </div>
      <div class="mg-input__input-container">
        <div class="mg-input__input">
          <select aria-describedby="" class="mg-input__box" id="identifier" name="identifier" title="Veuillez sélectionner une valeur">
            <option value="">
              Veuillez sélectionner une valeur
            </option>
            <option value="0">
              blu
            </option>
            <option value="1">
              bli
            </option>
            <option value="2">
              blo
            </option>
            <option value="3">
              bla
            </option>
          </select>
        </div>
      </div>
    </div>
  </mock:shadow-root>
</mg-input-select>
`;

exports[`mg-input-select Should render with args { label: 'label', identifier: 'identifier', items: [Array] }: 1`] = `
<mg-input-select>
  <mock:shadow-root>
    <div class="mg-input mg-input--select">
      <mg-input-title identifier="identifier">
        label
      </mg-input-title>
      <div class="mg-input__input-container">
        <div class="mg-input__input">
          <select aria-describedby="" class="mg-input__box" id="identifier" name="identifier" title="Veuillez sélectionner une valeur">
            <option value="">
              Veuillez sélectionner une valeur
            </option>
            <option value="0">
              blu
            </option>
            <option value="1">
              bli
            </option>
            <option value="2">
              blo
            </option>
            <option value="3">
              bla
            </option>
          </select>
        </div>
      </div>
    </div>
  </mock:shadow-root>
</mg-input-select>
`;

exports[`mg-input-select Should render with args { label: 'label', identifier: 'identifier', items: [Array] }: 2`] = `
<mg-input-select>
  <mock:shadow-root>
    <div class="mg-input mg-input--select">
      <mg-input-title identifier="identifier">
        label
      </mg-input-title>
      <div class="mg-input__input-container">
        <div class="mg-input__input">
          <select aria-describedby="" class="mg-input__box" id="identifier" name="identifier" title="Veuillez sélectionner une valeur">
            <option value="">
              Veuillez sélectionner une valeur
            </option>
            <option value="0">
              blu
            </option>
            <option value="1">
              bli
            </option>
            <option value="2">
              blo
            </option>
            <option value="3">
              bla
            </option>
          </select>
        </div>
      </div>
    </div>
  </mock:shadow-root>
</mg-input-select>
`;

exports[`mg-input-select Should render with args { label: 'label', identifier: 'identifier', items: [Array] }: 3`] = `
<mg-input-select>
  <mock:shadow-root>
    <div class="mg-input mg-input--select">
      <mg-input-title identifier="identifier">
        label
      </mg-input-title>
      <div class="mg-input__input-container">
        <div class="mg-input__input">
          <select aria-describedby="" class="mg-input__box" id="identifier" name="identifier" title="Veuillez sélectionner une valeur">
            <option value="">
              Veuillez sélectionner une valeur
            </option>
            <optgroup label="blu">
              <option value="0">
                blu
              </option>
              <option value="1">
                bli
              </option>
            </optgroup>
            <optgroup label="blo">
              <option value="0">
                blo
              </option>
            </optgroup>
            <option value="2">
              bla
            </option>
          </select>
        </div>
      </div>
    </div>
  </mock:shadow-root>
</mg-input-select>
`;

exports[`mg-input-select Should trigger events for items ({ items: [Array], selectedIndex: 0 }) with selectedIndex (%s) 1`] = `
<mg-input-select>
  <mock:shadow-root>
    <div class="mg-input mg-input--select">
      <mg-input-title identifier="identifier">
        label
      </mg-input-title>
      <div class="mg-input__input-container">
        <div class="mg-input__input">
          <select aria-describedby="identifier-help-text" class="mg-input__box" id="identifier" name="identifier" title="Veuillez sélectionner une valeur">
            <option value="">
              Veuillez sélectionner une valeur
            </option>
            <option value="0">
              batman
            </option>
            <option value="1">
              robin
            </option>
          </select>
        </div>
        <div class="mg-input__help-text" id="identifier-help-text">
          My help text
        </div>
      </div>
    </div>
  </mock:shadow-root>
</mg-input-select>
`;

exports[`mg-input-select Should trigger events for items ({ items: [Array], selectedIndex: 1 }) with selectedIndex (%s) 1`] = `
<mg-input-select>
  <mock:shadow-root>
    <div class="mg-input mg-input--select">
      <mg-input-title identifier="identifier">
        label
      </mg-input-title>
      <div class="mg-input__input-container">
        <div class="mg-input__input">
          <select aria-describedby="identifier-help-text" class="mg-input__box" id="identifier" name="identifier" title="Veuillez sélectionner une valeur">
            <option value="">
              Veuillez sélectionner une valeur
            </option>
            <option value="0">
              batman
            </option>
            <option value="1">
              robin
            </option>
            <option value="2">
              joker
            </option>
            <option value="3">
              bane
            </option>
          </select>
        </div>
        <div class="mg-input__help-text" id="identifier-help-text">
          My help text
        </div>
      </div>
    </div>
  </mock:shadow-root>
</mg-input-select>
`;

exports[`mg-input-select Should trigger events for items ({ items: [Array], selectedIndex: 1 }) with selectedIndex (%s) 2`] = `
<mg-input-select>
  <mock:shadow-root>
    <div class="mg-input mg-input--select">
      <mg-input-title identifier="identifier">
        label
      </mg-input-title>
      <div class="mg-input__input-container">
        <div class="mg-input__input">
          <select aria-describedby="identifier-help-text" class="mg-input__box" id="identifier" name="identifier" title="Veuillez sélectionner une valeur">
            <option value="">
              Veuillez sélectionner une valeur
            </option>
            <option value="0">
              batman
            </option>
            <option value="1">
              robin
            </option>
          </select>
        </div>
        <div class="mg-input__help-text" id="identifier-help-text">
          My help text
        </div>
      </div>
    </div>
  </mock:shadow-root>
</mg-input-select>
`;

exports[`mg-input-select Should trigger events for items ({ items: [Array], selectedIndex: 2 }) with selectedIndex (%s) 1`] = `
<mg-input-select>
  <mock:shadow-root>
    <div class="mg-input mg-input--select">
      <mg-input-title identifier="identifier">
        label
      </mg-input-title>
      <div class="mg-input__input-container">
        <div class="mg-input__input">
          <select aria-describedby="identifier-help-text" class="mg-input__box" id="identifier" name="identifier" title="Veuillez sélectionner une valeur">
            <option value="">
              Veuillez sélectionner une valeur
            </option>
            <option value="0">
              batman
            </option>
            <option value="1">
              robin
            </option>
            <option value="2">
              joker
            </option>
            <option value="3">
              bane
            </option>
          </select>
        </div>
        <div class="mg-input__help-text" id="identifier-help-text">
          My help text
        </div>
      </div>
    </div>
  </mock:shadow-root>
</mg-input-select>
`;

exports[`mg-input-select Should trigger events for items ({ items: [Array], selectedIndex: 3 }) with selectedIndex (%s) 1`] = `
<mg-input-select>
  <mock:shadow-root>
    <div class="mg-input mg-input--select">
      <mg-input-title identifier="identifier">
        label
      </mg-input-title>
      <div class="mg-input__input-container">
        <div class="mg-input__input">
          <select aria-describedby="identifier-help-text" class="mg-input__box" id="identifier" name="identifier" title="Veuillez sélectionner une valeur">
            <option value="">
              Veuillez sélectionner une valeur
            </option>
            <option value="0">
              batman
            </option>
            <option value="1">
              robin
            </option>
            <option value="2">
              joker
            </option>
            <option value="3">
              bane
            </option>
          </select>
        </div>
        <div class="mg-input__help-text" id="identifier-help-text">
          My help text
        </div>
      </div>
    </div>
  </mock:shadow-root>
</mg-input-select>
`;

exports[`mg-input-select validity, case next state is disabled validity ({ validity: false, valueMissing: true }), valueMissing (%s) 1`] = `
<mg-input-select patternerrormessage="Non">
  <mock:shadow-root>
    <div class="is-not-valid mg-input mg-input--select">
      <mg-input-title identifier="identifier">
        label
      </mg-input-title>
      <div class="mg-input__input-container">
        <div class="mg-input__input">
          <select aria-describedby="identifier-error" class="mg-input__box" id="identifier" name="identifier" title="Veuillez sélectionner une valeur">
            <option value="">
              Veuillez sélectionner une valeur
            </option>
            <option value="blu">
              blu
            </option>
            <option value="bli">
              bli
            </option>
            <option value="blo">
              blo
            </option>
            <option value="bla">
              bla
            </option>
          </select>
        </div>
        <div aria-live="assertive" class="mg-input__error" id="identifier-error">
          Ce champ est obligatoire.
        </div>
      </div>
    </div>
  </mock:shadow-root>
</mg-input-select>
`;

exports[`mg-input-select validity, case next state is disabled validity ({ validity: false, valueMissing: true }), valueMissing (%s) 2`] = `
<mg-input-select patternerrormessage="Non">
  <mock:shadow-root>
    <div class="mg-input mg-input--select">
      <mg-input-title identifier="identifier">
        label
      </mg-input-title>
      <div class="mg-input__input-container">
        <div class="mg-input__input">
          <select aria-describedby="identifier-error" class="mg-input__box" disabled="" id="identifier" name="identifier" title="Veuillez sélectionner une valeur">
            <option value="">
              Veuillez sélectionner une valeur
            </option>
            <option value="blu">
              blu
            </option>
            <option value="bli">
              bli
            </option>
            <option value="blo">
              blo
            </option>
            <option value="bla">
              bla
            </option>
          </select>
        </div>
        <div aria-live="assertive" class="mg-input__error" id="identifier-error">
          Ce champ est obligatoire.
        </div>
      </div>
    </div>
  </mock:shadow-root>
</mg-input-select>
`;

exports[`mg-input-select validity, case next state is readonly validity ({ validity: false, valueMissing: true }), valueMissing (%s) 1`] = `
<mg-input-select patternerrormessage="Non">
  <mock:shadow-root>
    <div class="is-not-valid mg-input mg-input--select">
      <mg-input-title identifier="identifier">
        label
      </mg-input-title>
      <div class="mg-input__input-container">
        <div class="mg-input__input">
          <select aria-describedby="identifier-error" class="mg-input__box" id="identifier" name="identifier" title="Veuillez sélectionner une valeur">
            <option value="">
              Veuillez sélectionner une valeur
            </option>
            <option value="blu">
              blu
            </option>
            <option value="bli">
              bli
            </option>
            <option value="blo">
              blo
            </option>
            <option value="bla">
              bla
            </option>
          </select>
        </div>
        <div aria-live="assertive" class="mg-input__error" id="identifier-error">
          Ce champ est obligatoire.
        </div>
      </div>
    </div>
  </mock:shadow-root>
</mg-input-select>
`;

exports[`mg-input-select validity, case next state is readonly validity ({ validity: false, valueMissing: true }), valueMissing (%s) 2`] = `
<mg-input-select patternerrormessage="Non">
  <mock:shadow-root>
    <div class="mg-input mg-input--readonly mg-input--select">
      <mg-input-title identifier="identifier">
        label
      </mg-input-title>
      <div class="mg-input__input-container">
        <strong></strong>
      </div>
    </div>
  </mock:shadow-root>
</mg-input-select>
`;
